cleanup(modern_bpf): r/w ro ebpf maps best practices

LucaGuerra opened the pull request:
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind cleanup

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

/area driver-modern-bpf

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**Does this PR require a change in the driver versions?**

No

<!--
Please remove the leading whitespace before the `/version <>` you uncommented.
-->

**What this PR does / why we need it**:

Our ebpf probe is saving many things, including the interesting syscall set, into the `.bss` section which is set by libbpf as mmapable. It is a best practice not to include security sensitive settings in an mmapable fd, so we are separating these, plus the general ebpf probe settings into their own maps; in addition, we are setting variables that are never written to read-only (by using `.rodata`).

Credits to @mouadk for looking at the maps permissions and suggesting this improvement!

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, you have to do nothing.
If yes, a release note is required:
Delete `NONE` and enter your extended release note in the block below.
Please note, the release note follows the "conventional commit specification" (https://www.conventionalcommits.org/en/v1.0.0/):
For example: `fix: broken link`.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
cleanup(modern_bpf): r/w ro ebpf maps best practices
```

LucaGuerra commented:
@FedeDP : if you agree, I would propose this for version 0.21.0, WDYT?

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/2399#" title="Author self-approved">LucaGuerra</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [LucaGuerra]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

LucaGuerra commented:
/milestone 0.21.0

FedeDP commented:
Sure Luca!
/milestone 8.1.0+driver

FedeDP commented:
Will run kernel-testing framework against this PR and post the results! :pray:

FedeDP commented:
Also would love a review by @Andreagit97 :heart:

FedeDP commented:
|        KERNEL        | CMAKE-CONFIGURE | KMOD BUILD | KMOD SCAP-OPEN |            BPF-PROBE BUILD             |            BPF-PROBE SCAP-OPEN             |            MODERN-BPF SCAP-OPEN             |
|----------------------|-----------------|------------|----------------|----------------------------------------|--------------------------------------------|---------------------------------------------|
| amazonlinux2-5.4     | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | [🟡](#amazonlinux2-54-modern-bpf_scap-open) |
| amazonlinux2022-5.15 | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                          |
| fedora-6.2           | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                          |
| oraclelinux-4.14     | 🟢              | 🟢         | 🟢             | [🟡](#oraclelinux-414-bpf-probe_build) | [🟡](#oraclelinux-414-bpf-probe_scap-open) | [🟡](#oraclelinux-414-modern-bpf_scap-open) |
| oraclelinux-5.15     | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                          |
| ubuntu-6.5           | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                          |


|        KERNEL        | CMAKE-CONFIGURE | KMOD BUILD | KMOD SCAP-OPEN |            BPF-PROBE BUILD             |            BPF-PROBE SCAP-OPEN             |             MODERN-BPF SCAP-OPEN             |
|----------------------|-----------------|------------|----------------|----------------------------------------|--------------------------------------------|----------------------------------------------|
| amazonlinux2-4.19    | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | [🟡](#amazonlinux2-419-modern-bpf_scap-open) |
| amazonlinux2-5.10    | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                           |
| amazonlinux2-5.15    | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                           |
| amazonlinux2-5.4     | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | [🟡](#amazonlinux2-54-modern-bpf_scap-open)  |
| amazonlinux2022-5.15 | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                           |
| amazonlinux2023-6.1  | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                           |
| archlinux-6.0        | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                           |
| archlinux-6.7        | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                           |
| centos-3.10          | 🟢              | 🟢         | 🟢             | [🟡](#centos-310-bpf-probe_build)      | [🟡](#centos-310-bpf-probe_scap-open)      | [🟡](#centos-310-modern-bpf_scap-open)       |
| centos-4.18          | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                           |
| centos-5.14          | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                           |
| fedora-5.17          | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                           |
| fedora-5.8           | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                           |
| fedora-6.2           | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                           |
| oraclelinux-3.10     | 🟢              | 🟢         | 🟢             | [🟡](#oraclelinux-310-bpf-probe_build) | [🟡](#oraclelinux-310-bpf-probe_scap-open) | [🟡](#oraclelinux-310-modern-bpf_scap-open)  |
| oraclelinux-4.14     | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | [🟡](#oraclelinux-414-modern-bpf_scap-open)  |
| oraclelinux-5.15     | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                           |
| oraclelinux-5.4      | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | [🟡](#oraclelinux-54-modern-bpf_scap-open)   |
| ubuntu-4.15          | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | [🟡](#ubuntu-415-modern-bpf_scap-open)       |
| ubuntu-5.8           | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | [🟡](#ubuntu-58-modern-bpf_scap-open)        |
| ubuntu-6.5           | 🟢              | 🟢         | 🟢             | 🟢                                     | 🟢                                         | 🟢                                           |

Andreagit97 commented on userspace/libpman/src/maps.c:
```diff
@@ -75,76 +75,159 @@ uint64_t pman_get_probe_schema_ver() {
+
 void pman_set_snaplen(uint32_t desired_snaplen) {
-	g_state.skel->bss->g_settings.snaplen = desired_snaplen;
+	struct capture_settings settings;
+	pman_get_capture_settings(&settings);
```
I would return here if we are not able to get the settings. There is the risk we will write junk at the next `update`
```suggestion
	struct capture_settings settings = { 0 };
	if(pman_get_capture_settings(&settings)!=0){return;}
```

LucaGuerra replied:
Makes sense, done!

Andreagit97 commented on userspace/libpman/src/maps.c:
```diff
@@ -75,76 +75,159 @@ uint64_t pman_get_probe_schema_ver() {
 	}
 }
 
+int pman_init_settings_map() {
```
it sounds like `pman_update_capture_settings` with an initial `struct capture_settings settings = {};` maybe we can avoid duplication

LucaGuerra replied:
Good catch, implemented your suggestion

Andreagit97 commented on driver/modern_bpf/helpers/base/maps_getters.h:
```diff
@@ -89,8 +144,12 @@ static __always_inline uint8_t maps__64bit_sampling_syscall_table(uint32_t sysca
-static __always_inline bool maps__64bit_interesting_syscall(uint32_t syscall_id) {
-	return g_64bit_interesting_syscalls_table[syscall_id & (SYSCALL_TABLE_SIZE - 1)];
+static __always_inline bool maps__interesting_syscall_64bit(uint32_t syscall_id) {
+	bool *ret = bpf_map_lookup_elem(&interesting_syscalls_table_64bit, &syscall_id);
```
This is the main reason we chose a memory-mapped global variable initially. Calling a bpf helper many times (more than 20 Millions/s) could decrease perf. We are probably talking about a few ns, but under heavy loads, it could have a little impact

FedeDP replied:
As far as we get it, `BPF_F_MMAPABLE` flag mmaps the eBPF maps to userspace memory, increasing perf for userspace only; in eBPF, the read and writes still pass through the helpers.
Perhaps we are wrong though?
See: 
https://docs.kernel.org/bpf/map_array.html
> Since kernel 5.5, memory mapping may be enabled for BPF_MAP_TYPE_ARRAY by setting the flag BPF_F_MMAPABLE. The map definition is page-aligned and starts on the first page. Sufficient page-sized and page-aligned blocks of memory are allocated to store all array values, starting on the second page, which in some cases will result in over-allocation of memory. The benefit of using this is increased performance and ease of use since userspace programs would not be required to use helper functions to access and mutate data.

https://docs.ebpf.io/linux/map-type/BPF_MAP_TYPE_ARRAY/#bpf_f_mmapable
> Setting this flag on a BPF_MAP_TYPE_ARRAY will allow userspace programs to [mmap](https://man7.org/linux/man-pages/man2/mmap.2.html) the array values into the userspace process, effectively making a shared memory region between eBPF programs and a userspace program.
This can significantly improve read and write performance since there is no syscall overhead to access the map.

https://patchwork.ozlabs.org/project/netdev/patch/20191117172806.2195367-4-andriin@fb.com/
> Add ability to memory-map contents of BPF array map. This is extremely useful
for working with BPF global data from userspace programs. It allows to avoid
typical bpf_map_{lookup,update}_elem operations, improving both performance
and usability.

Andreagit97 replied:
Uhm if I recall it correctly, the global variables are accessed without helpers but just using direct memory access. I quickly tried it 
```c
char LICENSE[] SEC("license") = "Dual BSD/GPL";

unsigned long stop = 0;

SEC("tp_btf/sys_enter")
int global(void *ctx) {
	stop++;
	return 0;
}

struct {
	__uint(type, BPF_MAP_TYPE_ARRAY);
	__uint(max_entries, 1);
	__type(key, uint32_t);
	__type(value, unsigned long);
} stop_map SEC(".maps");

SEC("tp_btf/sys_enter")
int map_array(void *ctx) {
	uint32_t key = 0;
	unsigned long *value = bpf_map_lookup_elem(&stop_map, &key);
	*value++;
	return 0;
}
```

Globals:

```
0: R1=ctx() R10=fp0
; stop++; @ globals.bpf.c:12
0: (18) r1 = 0xffffbfb7c1226000       ; R1_w=map_value(map=globals_.bss,ks=4,vs=8)
2: (79) r2 = *(u64 *)(r1 +0)          ; R1_w=map_value(map=globals_.bss,ks=4,vs=8) R2_w=scalar()
3: (07) r2 += 1                       ; R2_w=scalar()
4: (7b) *(u64 *)(r1 +0) = r2          ; R1_w=map_value(map=globals_.bss,ks=4,vs=8) R2_w=scalar()
; return 0; @ globals.bpf.c:13
5: (b7) r0 = 0                        ; R0_w=0
6: (95) exit
processed 6 insns (limit 1000000) max_states_per_insn 0 total_states 0 peak_states 0 mark_read 0
```

Map:

```
0: R1=ctx() R10=fp0
; int map_array(void *ctx) { @ globals.bpf.c:24
0: (b7) r1 = 0                        ; R1_w=0
; uint32_t key = 0; @ globals.bpf.c:25
1: (63) *(u32 *)(r10 -4) = r1
mark_precise: frame0: last_idx 1 first_idx 0 subseq_idx -1 
mark_precise: frame0: regs=r1 stack= before 0: (b7) r1 = 0
2: R1_w=0 R10=fp0 fp-8=0000????
2: (bf) r2 = r10                      ; R2_w=fp0 R10=fp0
;  @ globals.bpf.c:0
3: (07) r2 += -4                      ; R2_w=fp-4
; unsigned long *value = bpf_map_lookup_elem(&stop_map, &key); @ globals.bpf.c:26
4: (18) r1 = 0xffff9a8a4296b400       ; R1_w=map_ptr(map=stop_map,ks=4,vs=8)
6: (85) call bpf_map_lookup_elem#1    ; R0_w=map_value_or_null(id=1,map=stop_map,ks=4,vs=8)
; return 0; @ globals.bpf.c:28
7: (b7) r0 = 0                        ; R0_w=0
8: (95) exit
processed 8 insns (limit 1000000) max_states_per_insn 0 total_states 0 peak_states 0 mark_read 0
```

Probably looking at the bytecode of our probe before/after we should see the same

FedeDP replied:
Cool thanks for clarifying it! I guess you are right then, the emitted code is more optimized with the globals.

LucaGuerra replied:
Hi @Andreagit97 , makes sense! Indeed, the two compiled programs look different:

Using a map:
```
0000000000000090 <LBB0_26>:
      18:       63 7a f8 ff 00 00 00 00 *(u32 *)(r10 - 0x8) = r7
      19:       bf a2 00 00 00 00 00 00 r2 = r10
      20:       07 02 00 00 f8 ff ff ff r2 += -0x8
      21:       18 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r1 = 0x0 ll
      23:       85 00 00 00 01 00 00 00 call 0x1
```

Using `.bss`:
```
      18:       bf 81 00 00 00 00 00 00 r1 = r8
      19:       57 01 00 00 ff 01 00 00 r1 &= 0x1ff
      20:       18 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r2 = 0x0 ll
      22:       0f 12 00 00 00 00 00 00 r2 += r1
      23:       71 22 00 00 00 00 00 00 r2 = *(u8 *)(r2 + 0x0)
```

In the compiled object file. There is obviously a call. However, at load time the kernel is able to optimize this since it's an (almost)fixed-size array map; this uses more instructions than the `.bss` version but I guess it's not as bad as an actual call, let's take a look:

```
  17: (18) r1 = map[id:8834]
       18 11 00 00 82 22 00 00 00 00 00 00 00 00 00 00
  19: (07) r1 += 272
       07 01 00 00 10 01 00 00
  20: (61) r0 = *(u32 *)(r2 +0)
       61 20 00 00 00 00 00 00
  21: (35) if r0 >= 0x200 goto pc+3
       35 00 03 00 00 02 00 00
  22: (67) r0 <<= 3
       67 00 00 00 03 00 00 00
  23: (0f) r0 += r1
       0f 10 00 00 00 00 00 00
  24: (05) goto pc+1
       05 00 01 00 00 00 00 00
  25: (b7) r0 = 0
       b7 00 00 00 00 00 00 00
; if(ret == NULL) {
  26: (55) if r0 != 0x0 goto pc+30
       55 00 1e 00 00 00 00 00
...
; return *ret;
  57: (71) r1 = *(u8 *)(r0 +0)
       71 01 00 00 00 00 00 00
```

Originally we had 5 instructions, in the map version we have 10 instructions at the eBPF level. Not great but acceptable I think, also I'm not sure what happens at native code jit compile time or which kernel versions are able to perform this optimization.

Andreagit97 replied:
uhm, I see, I never noticed that! Yes, in this case, the overhead seems negligible

Andreagit97 commented on userspace/libpman/src/maps.c:
```diff
@@ -295,6 +378,38 @@ int pman_fill_syscall_exit_extra_tail_table() {
+int pman_mark_single_64bit_syscall(int syscall_id, bool interesting) {
+	char error_message[MAX_ERROR_MESSAGE_LEN];
+	int fd = bpf_map__fd(g_state.skel->maps.interesting_syscalls_table_64bit);
+	if(bpf_map_update_elem(fd, &syscall_id, &interesting, BPF_ANY) < 0) {
```
If this is called frequently it could have an impact as well but AFAIK we don't call it so frequently at runtime

Andreagit97 reviewed (commented):
Thank you for the changes! As i wrote in one comment the only downside that I could see is a possible decrease in perf.  We should probably keep an eye on possible perf regression after we merge this

LucaGuerra commented:
Thanks for review @Andreagit97 , addressed your comments. On the performance side we should keep an eye on it but it seems like we're only adding a few instructions and no calls if the loader is smart enough

poiana commented:
LGTM label has been added.  <details>Git tree hash: 839a6b6a14983b18d261a90ddd6e60c1b31ef040</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/2399#pullrequestreview-2835528387" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/2399#" title="Author self-approved">LucaGuerra</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,LucaGuerra]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/2399#pullrequestreview-2835528387" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/2399#pullrequestreview-2835667548" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/2399#" title="Author self-approved">LucaGuerra</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP,LucaGuerra]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP commented:
/unhold
